platform/android/display_server_android.cpp

```diff
@@ -74,6 +75,94 @@ String DisplayServerAndroid::get_name() const {
 	return "Android";
 }
 
+bool DisplayServerAndroid::tts_is_speaking() const {
+	GodotJavaWrapper *godot_java = OS_Android::get_singleton()->get_godot_java();
+	ERR_FAIL_COND_V(!godot_java, false);
+
+	return godot_java->tts_is_speaking();
+}
+
+bool DisplayServerAndroid::tts_is_paused() const {
+	GodotJavaWrapper *godot_java = OS_Android::get_singleton()->get_godot_java();
+	ERR_FAIL_COND_V(!godot_java, false);
+
+	return godot_java->tts_is_paused();
+}
+
+Array DisplayServerAndroid::tts_get_voices() const {
+	GodotJavaWrapper *godot_java = OS_Android::get_singleton()->get_godot_java();
+	ERR_FAIL_COND_V(!godot_java, Array());
+
+	return godot_java->tts_get_voices();
+}
+
+void DisplayServerAndroid::tts_speak(const String &p_text, const String &p_voice, int p_volume, float p_pitch, float p_rate, int p_utterance_id, bool p_interrupt) {
+	GodotJavaWrapper *godot_java = OS_Android::get_singleton()->get_godot_java();
+	ERR_FAIL_COND(!godot_java);
+	if (p_interrupt) {
+		tts_stop();
+	}
+
+	if (p_text.is_empty()) {
+		tts_post_utterance_event(DisplayServer::TTS_UTTERANCE_CANCELED, p_utterance_id);
+		return;
+	}
+
+	CharString string = p_text.utf8();
+	ids[p_utterance_id] = string;
+
+	godot_java->tts_speak(string.get_data(), p_voice.utf8().get_data(), CLAMP(p_volume, 0, 100), CLAMP(p_pitch, 0.f, 2.f), CLAMP(p_rate, 0.1f, 10.f), p_utterance_id, p_interrupt);
+}
+
+void DisplayServerAndroid::tts_pause() {
+	GodotJavaWrapper *godot_java = OS_Android::get_singleton()->get_godot_java();
+	ERR_FAIL_COND(!godot_java);
+
+	godot_java->tts_pause();
+}
+
+void DisplayServerAndroid::tts_resume() {
+	GodotJavaWrapper *godot_java = OS_Android::get_singleton()->get_godot_java();
+	ERR_FAIL_COND(!godot_java);
+
+	godot_java->tts_resume();
+}
+
+void DisplayServerAndroid::tts_stop() {
+	GodotJavaWrapper *godot_java = OS_Android::get_singleton()->get_godot_java();
+	ERR_FAIL_COND(!godot_java);
+	for (Map<int, CharString>::Element *E = ids.front(); E; E = E->next()) {
+		tts_post_utterance_event(DisplayServer::TTS_UTTERANCE_CANCELED, E->key());
+	}
+	ids.clear();
+
+	godot_java->tts_stop();
+}
+
+void DisplayServerAndroid::_java_utterance_callback(int p_event, int p_id, int p_pos) {
+	if (ids.has(p_id)) {
+		int pos = 0;
+		if ((TTSUtteranceEvent)p_event == DisplayServer::TTS_UTTERANCE_BOUNDARY) {
+			// Convert position from UTF-8 to UTF-32.
```

Comment:
Why is this needed?

Also, it may be better to add this utility method to `CharString` (if it's not available already) in order to prevent unnecessary code duplication.

Reply:
This should be UTF-16 not UTF-8, but the idea is the same, Java string and `Char16String` will have some characters (with codes > 0xFFFF) encoded as two codepoints, Godot `String` as one. This is to compensate the difference in offsets. Not sure if it's worth adding a utility method to the `CharString` / `Char16String`, UTF-8 version is used once for Javascript, and UTF-16 twice for Android and Windows, and unlikely more will be added.

Reply:
I think the fact it's used more than once warrant an addition :).

I'm mostly looking forward trying to avoid subtle bugs creeping in if one side is updated and the other not. With common methods, it's easy to see who's using it and the changes propagate automatically. 
With custom logic, you lose that advantage.